Assets/MRTK/Core/Providers/InputAnimation/InputAnimation.cs

```diff
@@ -462,272 +432,889 @@ private static int FindKeyframeInterval(AnimationCurve curve, float time)
+                outCurveX.AddKey(inCurveX[bestIndex]);
+                outCurveY.AddKey(inCurveY[bestIndex]);
+                outCurveZ.AddKey(inCurveZ[bestIndex]);
+                Recurse(start, bestIndex);
+                Recurse(bestIndex, end);
+            }
         }
 
-        private static void PoseCurvesToStream(BinaryWriter writer, PoseCurves curves, float startTime)
+        /// Utility function that creates a non-interpolated keyframe suitable for boolean values.
+        /// Keys are only added if the value changes.
+        /// Returns the index of the newly added keyframe, or -1 if no keyframe has been added.
+        [Obsolete]
+        private static int AddBoolKeyFiltered(AnimationCurve curve, float time, bool value)
         {
-            InputAnimationSerializationUtils.WriteFloatCurve(writer, curves.PositionX, startTime);
-            InputAnimationSerializationUtils.WriteFloatCurve(writer, curves.PositionY, startTime);
-            InputAnimationSerializationUtils.WriteFloatCurve(writer, curves.PositionZ, startTime);
+            float fvalue = value ? 1.0f : 0.0f;
+            // Set tangents and weights such than the input value is cut off and out tangent is constant.
+            var keyframe = new Keyframe(time, fvalue, 0.0f, 0.0f, 0.0f, boolOutWeight);
+            keyframe.weightedMode = WeightedMode.Both;
 
-            InputAnimationSerializationUtils.WriteFloatCurve(writer, curves.RotationX, startTime);
-            InputAnimationSerializationUtils.WriteFloatCurve(writer, curves.RotationY, startTime);
-            InputAnimationSerializationUtils.WriteFloatCurve(writer, curves.RotationZ, startTime);
-            InputAnimationSerializationUtils.WriteFloatCurve(writer, curves.RotationW, startTime);
+            int insertAfter = FindKeyframeInterval(curve, time);
+            if (insertAfter >= 0 && curve.keys[insertAfter].value == fvalue)
+            {
+                // Value unchanged from previous key, ignore
+                return -1;
+            }
+
+            int insertBefore = insertAfter + 1;
+            if (insertBefore < curve.keys.Length && curve.keys[insertBefore].value == fvalue)
+            {
+                // Same value as next key, replace next key
+                return curve.MoveKey(insertBefore, keyframe);
+            }
+
+            return curve.AddKey(keyframe);
         }
 
-        private static void PoseCurvesFromStream(BinaryReader reader, PoseCurves curves)
+        /// <summary>
+        /// Find an index i in the sorted events list, such that events[i].time &lt;= time &lt; events[i+1].time.
+        /// </summary>
+        /// <returns>
+        /// 0 &lt;= i &lt; eventCount if a full interval could be found.
+        /// -1 if time is less than the first event time.
+        /// eventCount-1 if time is greater than the last event time.
+        /// </returns>
+        /// <remarks>
+        /// Uses binary search.
+        /// </remarks>
+        [Obsolete]
```

Comment:
Would be good to add messages to these `Obsolete` tags, to describe what they're replaced with or (if no direct replacement) why they're marked as such.